valhalla/sif/costconstraints.h

```diff
@@ -0,0 +1,169 @@
+   * @param   edge        Pointer to a directed edge.
+   * @param   tile        Pointer to the tile which contains the directed edge for speed lookup
+   * @param   start       Time arriving at edge (secs)
+   * @param   end         Time leaving edge (secs)
+   * @param   is_forward  Boolean indicating whether this is a forward or reverse search
+   */
+  virtual void ComputeConstraintsForEdge(Cost& cost,
+                                         const baldr::DirectedEdge* edge,
+                                         const baldr::GraphTile* tile,
+                                         const uint32_t start,
+                                         const float end,
+                                         const bool is_forward) const {
+    auto ComputeConstraintForEdgeF = GetComputeConstraintForEdgeF();
+    if (constraints_.size() > 0)
+      for (uint32_t i = 0; i < constraints_.size(); ++i)
+        cost.data.push_back(ComputeConstraintForEdgeF(i, edge, tile, start, end, is_forward));
+  }
+
+  /**
+   * Compute all secondary constraints for the specified node.
+   * @param   cost        Cost where constraint calculations will be stored
+   * @param   id          Node Id
+   * @param   start       Time arriving at node (sec)
+   * @param   end         Time leaving node (float)
+   * @param   is_forward  Boolean indicating whether this is a forward or reverse search
+   */
+  virtual void ComputeConstraintsForNode(Cost& cost,
+                                         const baldr::GraphId& id,
+                                         const uint32_t start,
+                                         const float end,
+                                         const bool is_forward) const {
+    auto ComputeConstraintForNodeF = GetComputeConstraintForNodeF();
+    if (constraints_.size() > 0)
+      for (uint32_t i = 0; i < constraints_.size(); ++i)
+        cost.data.push_back(ComputeConstraintForNodeF(i, id, start, end, is_forward));
+  }
+  virtual inline bool ConstraintsSatisfied(const Cost& cost) const {
+    if (cost.data.size() == constraints_.size()) {
+      for (int i = 0; i < constraints_.size(); ++i) {
+        auto d = cost.data[i];
+        auto c = constraints_[i];
+        bool lower_satisfied = c.min_inclusive() ? d >= c.min_value() : d > c.min_value();
+        bool upper_satisfied = c.max_inclusive() ? d <= c.max_value() : d < c.max_value();
+        if (!lower_satisfied || !upper_satisfied)
+          return false;
+      }
+    }
+    return true;
+  }
+  inline bool HasConstraints() const {
+    return constraints_.size() > 0;
+  }
+
+protected:
+  std::vector<CostingConstraint> constraints_;
+  std::function<double(uint32_t,
+                       const baldr::DirectedEdge*,
+                       const baldr::GraphTile*,
+                       const uint32_t,
+                       const float,
+                       const bool)>
+  GetComputeConstraintForEdgeF() const {
+    if (ComputeConstraintForEdgeF != 0)
```

Comment:
Nit, shouldn't the comparison be to a bool? https://en.cppreference.com/w/cpp/utility/functional/function/operator_bool

Reply:
Yes.  I have spent most of my development time writing code in C#/Fortran/Python over the last few years, so I'm relearning what bit of C++ I picked up during my masters a handful of years ago.